### ISSUE TITLE
False positive for TS2729 (use before initialization) with target = ESNext

### ISSUE THREAD

**Issue by slavafomin** [microsoft/TypeScript#44298]
# Bug Report

Hello!

I've encountered this problem after updating my project to the latest TypeScript 4.3.2.

The bug looks similar to #37573.

#### script.ts

```ts
class Foo {

  constructor(private value) {
  }

  anotherValue = this.value;

}
```
#### tsconfig.json

```json
{
  "compilerOptions": {
    "target": "ESNext"
  }
}
```

This happens only when using `ESNext` as a `target`. Using `ES2020` works correctly. Could you also elaborate on the difference of these two settings?

### 🙁 Actual behavior

I'm getting an error: `error TS2729: Property 'value' is used before its initialization.`

### 🙂 Expected behavior

The code should compile just fine.

**Comment by whzx5byb** [comment]
My project also use `esnext` as the target but I can not reproduce the error.

**Comment by entropitor** [comment]
I have the same issue as @slavafomin when I upgraded to 4.3 in a project targetting esnext, es2020 solves it indeed!

**Comment by whzx5byb** [comment]
After some investigation, I think this is an intended behavior. 
Consider the following code:
```ts
class A {
  foo = this.bar;
  constructor(private bar: any) {}
}
```


A `esnext` target will compile it to the [native class field](https://github.com/tc39/proposal-class-fields).
And this is actually an error because `bar` is assigned to `foo` when it is not initialized.
```ts
// target: esnext
class A {
    bar;
    foo = this.bar;
    constructor(bar) {
        this.bar = bar;
    }
}
```


The `es2020` target will use `Object.defineProperty` to compile the code, and rearrange the assignment order.
```ts
class A {
    constructor(bar) {
        Object.defineProperty(this, "bar", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: bar
        });
        Object.defineProperty(this, "foo", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.bar
        });
    }
}
```
=========
PS:
> My project also use `esnext` as the target but I can not reproduce the error.

The reason is that I am using  `"useDefineForClassFields": false` in my project. It will produce the same output regardless of what the target is.
```ts
class A {
    constructor(bar) {
        this.bar = bar;
        this.foo = this.bar;
    }
}
```

**Comment by RyanCavanaugh** [comment]
@whzx5byb has it exactly right. The error is being issued exactly when a problem will occur.

**Comment by sirreal** [comment]
This seems to be the PR that introduced the (expected) change in compiler behavior: https://github.com/microsoft/TypeScript/pull/43194

**Comment by Bilge** [comment]
> The es2020 target will use Object.defineProperty to compile the code, and rearrange the assignment order.

Except that, in your example, the assignment order is still exactly the same.

**Comment by Artoria2e5** [comment]
Can reproduce with playground, although now with es2020 the transpile does not have Object.defineProperty:
```js
class Foo {
    constructor(value) {
        this.value = value;
        this.anotherValue = this.value;
    }
}
export {};
```

Bumped into this issue in https://github.com/homebridge/homebridge-plugin-template/blob/fe5bb73eee3efa007f137b9268e9a6c59b88d9ef/src/platform.ts#L12. There we have `api` being reported undefined.
